File: 8-string-to-integer-atoi/string-to-integer-atoi.cpp
class Solution {
public:
    int myAtoi(string s) {
        int64_t ans = 0;
        bool flag = false;
        bool begin = false;
        for (int i = 0; i < s.length(); i++) {
            if (s[i] == '+' || s[i] == '-') {
                if (begin) {
                    break;
                }
                begin = true;
                if(s[i] == '-') {
                    flag = true;
                }
                continue;
            }

            if (s[i] < '0' || s[i] > '9') {
                if ((isspace(s[i])) && !begin) {
                    continue;
                }
                break;
            }
            int num = s[i] - '0';
            ans = ans * 10 + num;
            if (ans > -static_cast<int64_t>(INT_MIN)) {
                break;
            }
            begin = true;
        }
        if (flag) {
            ans = -ans;
        }

         if (ans > INT_MAX) {
            ans = INT_MAX;
        }

        if (ans < INT_MIN) {
            ans = INT_MIN;
        }

        return static_cast<int>(ans);
    }
};

File: 8-string-to-integer-atoi/string-to-integer-atoi.go
func myAtoi(s string) int {
    sign := 1
    n := len(s)
    i := 0
    for ; i < n; i++ {
        if !unicode.IsSpace(rune(s[i])) {
            break
        }
    }
    if i < n && s[i] == '+' {
        sign = 1
        i++
    } else if i < n && s[i] == '-' {
        sign = -1
        i++
    }

    ans := 0
    for ; i < n; i++ {
        if s[i] < '0' || s[i] > '9' {
            break
        }
        v := s[i] - '0'
        ans = ans * 10 + int(v)
        if sign == 1 && ans > math.MaxInt32 {
            return int(math.MaxInt32)
        }
        if sign == -1 && (ans * sign) < math.MinInt32 {
            return int(math.MinInt32)
        } 
    }
    return ans * sign
}